modules/images/image-loading-optimization/storage/post-type.php

```diff
@@ -0,0 +1,191 @@
+			),
+			'public'           => false,
+			'hierarchical'     => false,
+			'rewrite'          => false,
+			'query_var'        => false,
+			'delete_with_user' => false,
+			'can_export'       => false,
+			'supports'         => array( 'title' ), // The original URL is stored in the post_title, and the post_name is a hash of the query vars.
+		)
+	);
+}
+add_action( 'init', 'ilo_register_page_metrics_post_type' );
+
+/**
+ * Get page metrics post.
+ *
+ * @param string $slug Page metrics slug.
+ * @return WP_Post|null Post object if exists.
+ */
+function ilo_get_page_metrics_post( $slug ) {
+	$post_query = new WP_Query(
+		array(
+			'post_type'              => ILO_PAGE_METRICS_POST_TYPE,
+			'post_status'            => 'publish',
+			'name'                   => $slug,
+			'posts_per_page'         => 1,
+			'no_found_rows'          => true,
+			'cache_results'          => true,
+			'update_post_meta_cache' => false,
+			'update_post_term_cache' => false,
+			'lazy_load_term_meta'    => false,
+		)
+	);
+
+	$post = current( $post_query->posts );
+	if ( $post instanceof WP_Post ) {
+		return $post;
+	} else {
+		return null;
+	}
+}
+
+/**
+ * Parses post content in page metrics post.
+ *
+ * @param WP_Post $post Page metrics post.
+ * @return array|WP_Error Page metrics when valid, or WP_Error otherwise.
+ */
+function ilo_parse_stored_page_metrics( WP_Post $post ) {
+	$page_metrics = json_decode( $post->post_content, true );
+	if ( json_last_error() ) {
+		return new WP_Error(
+			'page_metrics_json_parse_error',
+			sprintf(
+				/* translators: 1: Post type slug, 2: JSON error message */
+				__( 'Contents of %1$s post type not valid JSON: %2$s', 'performance-lab' ),
+				ILO_PAGE_METRICS_POST_TYPE,
+				json_last_error_msg()
+			)
+		);
+	}
+	if ( ! is_array( $page_metrics ) ) {
+		return new WP_Error(
+			'page_metrics_invalid_data_format',
+			sprintf(
+				/* translators: %s is post type slug */
+				__( 'Contents of %s post type was not a JSON array.', 'performance-lab' ),
+				ILO_PAGE_METRICS_POST_TYPE
+			)
+		);
+	}
+	return $page_metrics;
+}
+
+/**
+ * Gets page metrics for a slug.
+ *
+ * This is a convenience abstractions for lower-level functions.
+ *
+ * @see ilo_get_page_metrics_post()
+ * @see ilo_parse_stored_page_metrics()
+ *
+ * @param string $slug Page metrics slug.
+ * @return array Page metrics data, or empty array if invalid.
+ */
+function ilo_get_page_metrics_data( $slug ) {
+	$post = ilo_get_page_metrics_post( $slug );
+	if ( ! ( $post instanceof WP_Post ) ) {
+		return array();
+	}
+	$data = ilo_parse_stored_page_metrics( $post );
+	if ( ! is_array( $data ) ) {
+		return array();
+	}
+	return $data;
+}
+
+/**
+ * Stores page metric by merging it with the other page metrics for a given URL.
+ *
+ * The $validated_page_metric parameter has the following array shape:
+ *
+ * {
+ *      'viewport': array{
+ *          'width': int,
+ *          'height': int
+ *      },
+ *      'elements': array
+ * }
```

Comment:
WordPress doc standards have a way to describe array shapes within the parameter description. See https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#1-1-parameters-that-are-arrays

Reply:
Yeah, I was here originally trying to port the JSON schema over to an [array shape](https://phpstan.org/writing-php-code/phpdoc-types#array-shapes) in PHPStan for static analysis. WordPress's convention isn't supported. Since the array shape isn't needed for documentation here and since there is no type checking, I've just removed it in favor referencing the JSON Schema which does define it: a4ffbaef191b7d2b859c0a8bf08834fd57defedf

Reply:
That works too I suppose :-)

If it helps, what I usually do for these cases is follow WP‘s convention and then use a separate phpstan-param annotation for type checking.